### ISSUE TITLE
AutoMaterializePolicy not working for partitioned assets depending on a non-partitioned asset

### ISSUE THREAD

**Issue by rubenbriones** [dagster-io/dagster#15582]
### Dagster version

1.4.2

### What's the issue?

If an `AutoMaterializePolicy` is added to a partitioned asset, which depends on a non-partitioned asset, the partitions are only materialized automatically the first time (when they are in missing state), but then they are never materialized again, even if their parent asset is materialized. 

### How to reproduce?

1. Materialize `my_job` (2 partitions are added).
2. The asset `load_symbol` will be auto-materialized for the 2 created partitions.
3. Materialize `my_job` again.
4. The asset `load_symbol` will not be materialized,  even if their parent asset (`create_partitions`) have been just materialized.  

```python
from dagster import (
    asset,
    define_asset_job,
    AutoMaterializePolicy,
    DynamicPartitionsDefinition,
    Definitions,
)

symbols_partitions_def = DynamicPartitionsDefinition(name="symbols_partitions")


@asset
def download_zip(context) -> str:
    context.log.info("Downloading zip")
    path = "/home/file.zip"
    return path


@asset
def create_partitions(context, download_zip) -> None:
    context.log.info(f"Reading: {download_zip}")

    current_partitions = context.instance.get_dynamic_partitions("symbols_partitions")

    if len(current_partitions) == 0:
        new_partitions = ["EUR", "USD"]
    else:
        new_partitions = []

    if new_partitions:
        context.instance.add_dynamic_partitions("symbols_partitions", new_partitions)
        context.log.info(f"Added partitions: {new_partitions}")


@asset(
    deps=[create_partitions],
    partitions_def=symbols_partitions_def,
    auto_materialize_policy=AutoMaterializePolicy.eager(
        max_materializations_per_minute=10000
    ),
)
def load_symbol(context) -> None:
    context.log.info(f"Loading data for: {context.partition_ke}")


my_job = define_asset_job(
    name="my_job",
    selection=[download_zip, create_partitions],
)


defs = Definitions(
    assets=[download_zip, create_partitions, load_symbol],
    jobs=[my_job],
)
```

### Deployment type

None

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by rubenbriones** [comment]
Also, related to the example shown above, how can I limit the concurrent partitions executed in parallel? I would like to limit the concurrency for the `load_symbol` asset, and I have tried to do that by setting the `op_tags={"2_run_per_value": "load_symbol"}`, (and setting that tag in the `dagster.yaml` file inside the `tag_concurency_limits` section), but the tag is not being added to the asset materializations.

Before dagster 1.4, I had done that with something like that:

```python
sensor= build_asset_reconciliation_sensor(
    name="load_symbol_sensor",
    asset_selection=AssetSelection.assets(load_symbol),
    run_tags={"2_run_per_value": "load_symbol"},
)
```

But now, without the `build_asset_reconciliation_sensor` I don't know hoe to do it.

Thank you so much in advance for your help.

**Comment by rubenbriones** [comment]
I have just check that the bug was resolved in dagster 1.4.7, and now the auto-materizliaze policy is working OK.

I will move the ``run_tags`` question to a new issue, and close this.

**Comment by rubenbriones** [comment]
Sorry, I re-open the issue because I have seen that although in version 1.4.7 now it does re-run the partitioned asset that depends on the non-partitioned asset (when using the AutoMaterializePolicy.eager()), I am seeing that **it only re-executes 1 partition, instead of all the partitions** (in the example code it re-launches only the ``USD`` partition, and not the ``EUR`` partition).

In fact, you can see how even in the first run the 2 partitions are executed for being missing, but the "Upstream data has changed" condition is only detected OK for 1 partition.:

![image](https://github.com/dagster-io/dagster/assets/56679758/fd4e7c17-0e98-4d07-ba68-346dc597e915)
